[Page Color Sampler] Avoid detection/sampling unless fixed-position layers have been repainted or repositioned

whsieh opened the pull request:
#### 14b4507e839436794cc013d109cb22c0297528c6
<pre>
[Page Color Sampler] Avoid detection/sampling unless fixed-position layers have been repainted or repositioned
<a href="https://bugs.webkit.org/show_bug.cgi?id=290662">https://bugs.webkit.org/show_bug.cgi?id=290662</a>
<a href="https://rdar.apple.com/148135149">rdar://148135149</a>

Reviewed by Abrar Rahman Protyasha.

Refactor how fixed edge container detection/sampling works, such that it only engages after painting
or repositioning viewport-constrained layers. See below for more details.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setNeedsFixedContainerEdgesUpdate):

Add a new chrome client hook to call out into `WebPage`, to set the new flag.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::obscuredContentInsetsDidChange):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setObscuredInsets):
* Source/WebCore/page/Page.h:
(WebCore::Page::setObscuredInsets): Deleted.

Set the `m_needsFixedContainerEdgesUpdate` flag when insets change.

* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateGeometry):
(WebCore::RenderLayerBacking::setContentsNeedDisplay):
(WebCore::RenderLayerBacking::setContentsNeedDisplayInRect):
(WebCore::RenderLayerBacking::setNeedsFixedContainerEdgesUpdateIfNeeded):

Compute and send `FixedContainerEdges` after the next layer tree commit, if geometry of any
fixed/sticky layers changes or the layer needs to be repainted.

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:

Change `m_fixedContainerEdges` to be a `std::optional`; this allows us to only update the cached
UI-side `_fixedContainerEdges` in the case where the transaction contains something other than
`std::nullopt`.

(WebKit::RemoteLayerTreeTransaction::fixedContainerEdges const):
(WebKit::RemoteLayerTreeTransaction::setFixedContainerEdges):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):

Only update `WKWebView` if the transaction contains `FixedContainerEdges`.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setNeedsFixedContainerEdgesUpdate):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):

Change this to only compute and send `FixedContainerEdges` if `m_needsFixedContainerEdgesUpdate` is
set.

(WebKit::WebPage::didCommitLoad):

Reset `m_needsFixedContainerEdgesUpdate` to its initial state.

(WebKit::WebPage::didAddOrRemoveViewportConstrainedObjects):

Set `m_needsFixedContainerEdgesUpdate` whenever the viewport-constrained objects set changes.

* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/292931@main">https://commits.webkit.org/292931@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7305c997bd581fad54aac46405fb15a1d43c0da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97457 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102544 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99502 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25534 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74258 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31438 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13153 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88136 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIDevTools.MessagePassingFromPanelToBackground (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54603 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12921 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6019 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47428 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104564 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24536 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17893 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/data-url-shared.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83302 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24908 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82723 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20821 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27267 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4957 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18103 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29666 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24320 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27634 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25894 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -435,7 +435,8 @@
 {
-    [webView() _updateFixedContainerEdges:transaction.fixedContainerEdges()];
+    if (transaction.fixedContainerEdges())
+        [webView() _updateFixedContainerEdges:*transaction.fixedContainerEdges()];
```
nit:
```suggestion
    if (const auto& edges = transaction.fixedContainerEdges())
        [webView() _updateFixedContainerEdges:*edges];
```

whsieh replied:
Fixed! I opted for just `auto&`, since we _tend_ to omit the `const` in most other places in the codebase.

whsieh committed:
[Page Color Sampler] Avoid detection/sampling unless fixed-position layers have been repainted or repositioned
https://bugs.webkit.org/show_bug.cgi?id=290662
rdar://148135149

Reviewed by Abrar Rahman Protyasha.

Refactor how fixed edge container detection/sampling works, such that it only engages after painting
or repositioning viewport-constrained layers. See below for more details.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setNeedsFixedContainerEdgesUpdate):

Add a new chrome client hook to call out into `WebPage`, to set the new flag.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::obscuredContentInsetsDidChange):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setObscuredInsets):
* Source/WebCore/page/Page.h:
(WebCore::Page::setObscuredInsets): Deleted.

Set the `m_needsFixedContainerEdgesUpdate` flag when insets change.

* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateGeometry):
(WebCore::RenderLayerBacking::setContentsNeedDisplay):
(WebCore::RenderLayerBacking::setContentsNeedDisplayInRect):
(WebCore::RenderLayerBacking::setNeedsFixedContainerEdgesUpdateIfNeeded):

Compute and send `FixedContainerEdges` after the next layer tree commit, if geometry of any
fixed/sticky layers changes or the layer needs to be repainted.

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:

Change `m_fixedContainerEdges` to be a `std::optional`; this allows us to only update the cached
UI-side `_fixedContainerEdges` in the case where the transaction contains something other than
`std::nullopt`.

(WebKit::RemoteLayerTreeTransaction::fixedContainerEdges const):
(WebKit::RemoteLayerTreeTransaction::setFixedContainerEdges):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):

Only update `WKWebView` if the transaction contains `FixedContainerEdges`.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setNeedsFixedContainerEdgesUpdate):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):

Change this to only compute and send `FixedContainerEdges` if `m_needsFixedContainerEdgesUpdate` is
set.

(WebKit::WebPage::didCommitLoad):

Reset `m_needsFixedContainerEdgesUpdate` to its initial state.

(WebKit::WebPage::didAddOrRemoveViewportConstrainedObjects):

Set `m_needsFixedContainerEdgesUpdate` whenever the viewport-constrained objects set changes.

* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/292931@main
